File: scripts/make_synthese_docx.py
#!/usr/bin/env python
"""Génère MS-VICRegL_synthese.docx — synthèse problématiques + résultats (python-docx)."""
from pathlib import Path

from docx import Document
from docx.enum.section import WD_ORIENT
from docx.enum.table import WD_TABLE_ALIGNMENT
from docx.enum.text import WD_ALIGN_PARAGRAPH
from docx.oxml.ns import qn
from docx.oxml import OxmlElement
from docx.shared import Pt, RGBColor, Inches

BLUE = RGBColor(0x29, 0x52, 0xCC)
RED = RGBColor(0xC0, 0x39, 0x2B)
GREEN = RGBColor(0x1E, 0x88, 0x49)
ORANGE = RGBColor(0xB9, 0x77, 0x0E)
GREY = RGBColor(0x5D, 0x6D, 0x7E)
DARK = RGBColor(0x1C, 0x28, 0x33)
WHITE = RGBColor(0xFF, 0xFF, 0xFF)

doc = Document()

# ---- base style ----
st = doc.styles["Normal"]
st.font.name = "Calibri"
st.font.size = Pt(10.5)
st.font.color.rgb = DARK

for name, sz, col in (("Heading 1", 15, DARK), ("Heading 2", 12, BLUE)):
    h = doc.styles[name]
    h.font.name = "Calibri"
    h.font.size = Pt(sz)
    h.font.bold = True
    h.font.color.rgb = col

sec = doc.sections[0]
sec.top_margin = sec.bottom_margin = Inches(0.75)
sec.left_margin = sec.right_margin = Inches(0.83)


def shade(cell, hex_fill):
    tcPr = cell._tc.get_or_add_tcPr()
    sh = OxmlElement("w:shd")
    sh.set(qn("w:val"), "clear")
    sh.set(qn("w:color"), "auto")
    sh.set(qn("w:fill"), hex_fill)
    tcPr.append(sh)


def run(p, text, *, bold=False, italic=False, color=None, size=None, mono=False):
    r = p.add_run(text)
    r.bold = bold
    r.italic = italic
    if color is not None:
        r.font.color.rgb = color
    if size is not None:
        r.font.size = Pt(size)
    if mono:
        r.font.name = "Consolas"
    return r


def para(runs, *, align=None, after=6, line=1.15):
    p = doc.add_paragraph()
    if align is not None:
        p.alignment = align
    pf = p.paragraph_format
    pf.space_after = Pt(after)
    pf.line_spacing = line
    for spec in runs:
        run(p, spec[0], **spec[1])
    return p


def bullet(runs):
    p = doc.add_paragraph(style="List Bullet")
    p.paragraph_format.space_after = Pt(3)
    for spec in runs:
        run(p, spec[0], **spec[1])
    return p


def r(text, **kw):
    return (text, kw)


def add_table(header, rows, widths):
    """rows: list of (cells, colors_list_or_None, bold_bool)."""
    tbl = doc.add_table(rows=1, cols=len(header))
    tbl.alignment = WD_TABLE_ALIGNMENT.CENTER
    tbl.autofit = False
    # header
    for i, (txt, w) in enumerate(zip(header, widths)):
        c = tbl.rows[0].cells[i]
        c.width = Inches(w)
        shade(c, "1C2833")
        pc = c.paragraphs[0]
        pc.alignment = WD_ALIGN_PARAGRAPH.LEFT if i == 0 else WD_ALIGN_PARAGRAPH.CENTER
        for j, line in enumerate(txt.split("\n")):
            if j:
                pc = c.add_paragraph()
                pc.alignment = WD_ALIGN_PARAGRAPH.CENTER
            run(pc, line, bold=True, color=WHITE, size=9.5)
    # body
    for ridx, (cells, colors, bold) in enumerate(rows):
        row = tbl.add_row()
        for i, txt in enumerate(cells):
            c = row.cells[i]
            c.width = Inches(widths[i])
            shade(c, "F4F6F7" if ridx % 2 else "FFFFFF")
            pc = c.paragraphs[0]
            pc.alignment = WD_ALIGN_PARAGRAPH.LEFT if i == 0 else WD_ALIGN_PARAGRAPH.CENTER
            pc.paragraph_format.space_after = Pt(2)
            col = colors[i] if colors else DARK
            for j, line in enumerate(str(txt).split("\n")):
                if j:
                    pc = c.add_paragraph()
                    pc.alignment = WD_ALIGN_PARAGRAPH.CENTER
                run(pc, line, bold=bold, color=col, size=9.5)
    return tbl


def hr(p, top=False):
    pPr = p._p.get_or_add_pPr()
    bd = OxmlElement("w:pBdr")
    edge = OxmlElement("w:top" if top else "w:bottom")
    edge.set(qn("w:val"), "single"); edge.set(qn("w:sz"), "12" if not top else "8")
    edge.set(qn("w:space"), "6"); edge.set(qn("w:color"), "2952CC" if not top else "D5D8DC")
    bd.append(edge); pPr.append(bd)


# ===================== TITRE =====================
t = doc.add_paragraph(); t.alignment = WD_ALIGN_PARAGRAPH.CENTER
run(t, "Identification bactérienne par MALDI-TOF", bold=True, size=17, color=DARK)
s = doc.add_paragraph(); s.alignment = WD_ALIGN_PARAGRAPH.CENTER
run(s, "Un CNN 1D auto-supervisé (VICRegL) invariant aux artefacts de centre", size=13, color=BLUE)
s2 = doc.add_paragraph(); s2.alignment = WD_ALIGN_PARAGRAPH.CENTER
run(s2, "Synthèse des problématiques et des résultats", italic=True, size=11, color=GREY)
s3 = doc.add_paragraph(); s3.alignment = WD_ALIGN_PARAGRAPH.CENTER
run(s3, "Jeu DRIAMS (4 centres hospitaliers) · comparaison au pipeline classique MSclassifR",
    size=9.5, color=GREY)
hr(s3)

# ===================== 1. CONTEXTE =====================
doc.add_heading("1. Contexte et problématique", level=1)
para([r("L’identification bactérienne par spectrométrie MALDI-TOF est routinière, mais les "
        "modèles souffrent d’un "), r("effet de centre", bold=True),
      r(" : calibration m/z, ligne de base, gain du détecteur et matrice varient d’un hôpital "
        "et d’un appareil à l’autre. Un classifieur entraîné dans un centre "),
      r("se dégrade sur un autre", italic=True),
      r(", ce qui limite le déploiement multi-sites.")], align=WD_ALIGN_PARAGRAPH.JUSTIFY)
para([r("Le pipeline de référence (type "), r("MSclassifR", bold=True),
      r(") neutralise ces artefacts par un pré-traitement lourd fait main (stabilisation de "
        "variance, ondelettes, SNIP, normalisation TIC, binning 3 Da, correction de lot). "),
      r("Question de ce travail : ", bold=True, color=BLUE),
      r("peut-on plutôt "), r("apprendre l’invariance aux artefacts", bold=True),
      r(", à partir de spectres quasi-bruts, sans ce pré-traitement expert ?")],
     align=WD_ALIGN_PARAGRAPH.JUSTIFY)

# ===================== 2. APPROCHE =====================
doc.add_heading("2. Approche proposée", level=1)
para([r("Un encodeur "), r("CNN 1D (ResNet-1D)", bold=True),
      r(" est pré-entraîné en "), r("auto-supervision VICRegL", bold=True),
      r(" — sans étiquettes. Les augmentations "), r("simulent les artefacts MALDI", italic=True),
      r(" (déformation de calibration m/z, ligne de base, gain, bruit, perte de pics) : le "
        "réseau apprend à ignorer ces nuisances tout en préservant la structure discriminante "
        "des pics (biomarqueurs) via le critère local de VICRegL.")],
     align=WD_ALIGN_PARAGRAPH.JUSTIFY)
para([r("Entrée : ", bold=True),
      r("spectre brut ré-échantillonné (2000–20000 Da, 6000 points) + normalisation TIC "
        "seulement.  "),
      r("Évaluation : ", bold=True),
      r("sonde linéaire sur features gelées, comparée à un Random Forest sur "),
      r("binned_6000", mono=True, size=10),
      r(" (représentation DRIAMS standard, pré-traitement complet).")],
     align=WD_ALIGN_PARAGRAPH.JUSTIFY)
para([r("Données DRIAMS : ", bold=True),
      r("4 centres — B (Bâle-Land), C (Aarau), D (Viollier), A (Bâle). Étiquettes "),
      r("Biotyper (MALDI)", bold=True, color=ORANGE),
      r(". Métrique principale : "), r("balanced accuracy", italic=True),
      r(" (classes déséquilibrées).")], align=WD_ALIGN_PARAGRAPH.JUSTIFY)

# ===================== 3. RÉSULTATS =====================
doc.add_heading("3. Résultats — vue d’ensemble", level=1)
add_table(
    ["Expérience", "Résultat clé", "Verdict"],
    [
        (["ID espèces cross-centre B↔C",
          "VICRegL 0,95–0,99 (stable) vs RF 0,68–1,00 (erratique) — ~8× plus stable",
          "Favorable*"], [DARK, DARK, GREEN], False),
        (["Espèces proches (Enterobacter, Strepto, Klebsiella)",
          "VICRegL > RF sur les 3 groupes (McNemar p<0,01)", "Favorable*"],
         [DARK, DARK, GREEN], False),
        (["Résistance antibiotique (AMR)",
          "RF bat VICRegL, aucun gain de stabilité", "Défavorable"],
         [DARK, DARK, RED], False),
        (["Hold-out centre D (jamais vu)",
          "RF 0,935 > VICRegL 0,886 — renversement", "Défavorable"],
         [DARK, DARK, RED], False),
        (["D ajouté au pré-entraînement (non-labellisé)",
          "0,886 → 0,915 : écart au RF divisé par 2", "Partiel"],
         [DARK, DARK, ORANGE], False),
    ],
    [2.35, 3.25, 1.15],
)
para([r("* dans le cadre où les deux centres ont été vus (non-labellisés) au pré-entraînement "
        "— voir §4.", italic=True, size=9, color=GREY)], after=8)

# ===================== 4. POINT CRITIQUE =====================
doc.add_heading("4. Le point critique : adaptation ≠ généralisation", level=1)
para([r("Le résultat marquant « ~8× plus stable » (B↔C) est obtenu quand "),
      r("les deux centres ont été vus, non-labellisés, au pré-entraînement", bold=True),
      r(" : c’est une "), r("adaptation de domaine à cible disponible", italic=True),
      r(", pas de la généralisation à un centre neuf. Le "), r("test strict", bold=True),
      r(" — centre D tenu totalement à l’écart du SSL — "),
      r("inverse la conclusion", bold=True, color=RED),
      r(" : le pipeline classique l’emporte.")], align=WD_ALIGN_PARAGRAPH.JUSTIFY)
add_table(
    ["Condition (éval. sur D)", "Encodeur B∪C\n(D jamais vu)",
     "Encodeur B∪C∪D\n(D vu non-labellisé)", "RF classique"],
    [
        (["(B+C) → D  (centre inédit)", "0,886", "0,915", "0,935"],
         [DARK, RED, ORANGE, GREEN], True),
        (["Plafond D → D (sonde sur D)", "0,914", "0,931", "—"], None, False),
        (["B+C en intra-domaine (plafond)", "0,988", "0,990", "0,889"], None, False),
        (["AUROC « quel centre ? » sur features", "0,997", "0,999", "—"],
         [DARK, RED, RED, DARK], False),
    ],
    [2.55, 1.55, 1.75, 1.15],
)
para([r("Balanced accuracy ; 10 espèces communes B∩C∩D ; test = 8797 spectres de D.",
        italic=True, size=9, color=GREY)], after=8)
para([r("Diagnostic : ", bold=True, color=BLUE),
      r("l’AUROC « quel centre ? » reste ~1,0 même après avoir vu D — les features "),
      r("encodent encore parfaitement le centre", bold=True),
      r(". L’invariance revendiquée "), r("n’est pas atteinte", bold=True, color=RED),
      r(" ; ce qui fonctionnait en B↔C relevait de l’adaptation. Voir D en SSL enrichit la "
        "représentation (gain réel, sans labels) mais "),
      r("ne restaure ni la parité avec le RF ni l’invariance", italic=True), r(".")],
     align=WD_ALIGN_PARAGRAPH.JUSTIFY)

# ===================== 5. FORCES / LIMITES =====================
doc.add_heading("5. Forces et limites", level=1)
doc.add_heading("Forces", level=2)
bullet([r("Pas de pré-traitement expert", bold=True),
        r(" : spectre quasi-brut + TIC ; plafond intra-domaine élevé (0,99).")])
bullet([r("Extension d’un centre sans aucun label", bold=True),
        r(" : l’ajouter au SSL (non-supervisé) réduit de moitié l’écart au pipeline classique.")])
bullet([r("Diagnostic rigoureux", bold=True),
        r(" : décomposition plafond/transfert, classifieur de domaine, McNemar apparié, "
          "IC bootstrap.")])
doc.add_heading("Limites", level=2)
bullet([r("Circularité des étiquettes", bold=True, color=ORANGE),
        r(" : DRIAMS est labellisé Biotyper (MALDI), pas NGS — 0 Shigella dans B+C. On ne peut "
          "pas prétendre « battre le MALDI » avec des labels MALDI.")])
bullet([r("Invariance non atteinte", bold=True),
        r(" : AUROC domaine ~1,0 ; échec sur un centre vraiment inédit.")])
bullet([r("AMR", bold=True),
        r(" : les augmentations (perte de pics) effacent le signal faible que la résistance "
          "requiert.")])

# ===================== 6. CONCLUSION =====================
doc.add_heading("6. Conclusion et perspectives", level=1)
para([r("Le pipeline est un "), r("prototype de méthode prometteur", bold=True),
      r(", pas un système supérieur au classique : il excelle en intra-domaine et en "
        "adaptation, mais "), r("l’invariance apprise ne généralise pas encore", italic=True),
      r(" à un centre inédit, où le pré-traitement fait main reste devant.")],
     align=WD_ALIGN_PARAGRAPH.JUSTIFY)
para([r("Deux leviers pour la suite :", bold=True, color=BLUE)])
bullet([r("Terme d’invariance de centre explicite", bold=True),
        r(" dans la loss (domaine-adversarial / CORAL) — attaque directement l’AUROC ~1,0, "
          "cause racine de l’échec.")])
bullet([r("Données à étiquettes moléculaires (WGS)", bold=True),
        r(" — seul moyen de lever la circularité et de tester la vraie hypothèse : l’invariance "
          "capte-t-elle un signal que les labels MALDI ratent ?")])
para([r("Ce dernier point ouvre une "), r("collaboration naturelle", bold=True, color=GREEN),
      r(" avec les travaux sur données MALDI appariées à des labels moléculaires "
        "(identification fine de sous-espèces, rejet par confiance).")],
     align=WD_ALIGN_PARAGRAPH.JUSTIFY)
foot = doc.add_paragraph()
foot.paragraph_format.space_before = Pt(14)
hr(foot, top=True)
run(foot, "Reproductibilité : encodeur ResNet-1D, PyTorch MPS (Apple M3 Pro). Résultats "
          "régénérables via scripts/08_holdout_D.py et 09_ssl_BCD.py.",
    italic=True, size=8.5, color=GREY)

out = Path(__file__).resolve().parents[1] / "MS-VICRegL_synthese.docx"
doc.save(out)
print("OK ->", out)

File: scripts/make_synthese_docx.js
// Génère MS-VICRegL_synthese.docx — synthèse problématiques + résultats
const {
  Document, Packer, Paragraph, TextRun, HeadingLevel, AlignmentType,
  Table, TableRow, TableCell, WidthType, BorderStyle, ShadingType,
  LevelFormat, PageBreak
} = require('docx');
const fs = require('fs');

const BLUE = '2952cc', RED = 'c0392b', GREEN = '1e8449', ORANGE = 'b9770e';
const GREY = '5d6d7e', DARK = '1c2833';

// ---- helpers ----
const H1 = (t) => new Paragraph({ heading: HeadingLevel.HEADING_1, spacing: { before: 260, after: 120 }, children: [new TextRun({ text: t, bold: true, color: DARK })] });
const H2 = (t) => new Paragraph({ heading: HeadingLevel.HEADING_2, spacing: { before: 180, after: 80 }, children: [new TextRun({ text: t, bold: true, color: BLUE })] });
const P = (runs, opts = {}) => new Paragraph({ spacing: { after: 100, line: 276 }, alignment: AlignmentType.JUSTIFIED, children: Array.isArray(runs) ? runs : [new TextRun(runs)], ...opts });
const T = (text, o = {}) => new TextRun({ text, ...o });
const bullet = (runs) => new Paragraph({ numbering: { reference: 'bul', level: 0 }, spacing: { after: 60, line: 268 }, children: Array.isArray(runs) ? runs : [new TextRun(runs)] });

function shadeCell(text, { bg, bold, color, align, w }) {
  return new TableCell({
    width: { size: w, type: WidthType.DXA },
    shading: bg ? { type: ShadingType.CLEAR, color: 'auto', fill: bg } : undefined,
    margins: { top: 50, bottom: 50, left: 90, right: 90 },
    children: [new Paragraph({
      alignment: align || AlignmentType.LEFT,
      children: [new TextRun({ text, bold: !!bold, color: color || DARK, size: 19 })]
    })]
  });
}
function tableRow(cells, colW, { bg, bold, colors } = {}) {
  return new TableRow({
    children: cells.map((c, i) => shadeCell(c, {
      bg, bold, w: colW[i],
      color: colors ? colors[i] : undefined,
      align: i === 0 ? AlignmentType.LEFT : AlignmentType.CENTER
    }))
  });
}
function makeTable(header, rows, colW, rowOpts) {
  const total = colW.reduce((a, b) => a + b, 0);
  return new Table({
    columnWidths: colW,
    width: { size: total, type: WidthType.DXA },
    rows: [
      tableRow(header, colW, { bg: DARK, bold: true, colors: header.map(() => 'FFFFFF') }),
      ...rows.map((r, idx) => tableRow(r.cells, colW, {
        bg: idx % 2 ? 'F4F6F7' : 'FFFFFF', bold: r.bold, colors: r.colors
      }))
    ]
  });
}

const COL = { size: 9360, type: WidthType.DXA };

const doc = new Document({
  creator: 'MS-VICRegL',
  title: 'MS-VICRegL — Synthèse',
  numbering: {
    config: [{
      reference: 'bul', levels: [{
        level: 0, format: LevelFormat.BULLET, text: '•', alignment: AlignmentType.LEFT,
        style: { paragraph: { indent: { left: 360, hanging: 200 } } }
      }]
    }]
  },
  styles: {
    default: { document: { run: { font: 'Calibri', size: 21, color: DARK } } },
    paragraphStyles: [
      { id: 'Heading1', name: 'Heading 1', basedOn: 'Normal', next: 'Normal', quickFormat: true, run: { size: 28, bold: true, color: DARK }, paragraph: { spacing: { before: 260, after: 120 } } },
      { id: 'Heading2', name: 'Heading 2', basedOn: 'Normal', next: 'Normal', quickFormat: true, run: { size: 23, bold: true, color: BLUE }, paragraph: { spacing: { before: 180, after: 80 } } }
    ]
  },
  sections: [{
    properties: { page: { size: { width: 12240, height: 15840 }, margin: { top: 1080, bottom: 1080, left: 1200, right: 1200 } } },
    children: [
      // ===== TITRE =====
      new Paragraph({ alignment: AlignmentType.CENTER, spacing: { after: 40 }, children: [new TextRun({ text: 'Identification bactérienne par MALDI-TOF', bold: true, size: 34, color: DARK })] }),
      new Paragraph({ alignment: AlignmentType.CENTER, spacing: { after: 60 }, children: [new TextRun({ text: 'Un CNN 1D auto-supervisé (VICRegL) invariant aux artefacts de centre', size: 26, color: BLUE })] }),
      new Paragraph({ alignment: AlignmentType.CENTER, spacing: { after: 40 }, children: [new TextRun({ text: 'Synthèse des problématiques et des résultats', italics: true, size: 22, color: GREY })] }),
      new Paragraph({ alignment: AlignmentType.CENTER, spacing: { after: 200 }, border: { bottom: { color: BLUE, size: 12, style: BorderStyle.SINGLE, space: 8 } }, children: [new TextRun({ text: 'Jeu DRIAMS (4 centres hospitaliers) · comparaison au pipeline classique MSclassifR', size: 19, color: GREY })] }),

      // ===== 1. CONTEXTE =====
      H1('1. Contexte et problématique'),
      P([
        T('L’identification bactérienne par spectrométrie MALDI-TOF est routinière, mais les modèles souffrent d’un '),
        T('effet de centre', { bold: true }),
        T(' : calibration m/z, ligne de base, gain du détecteur et matrice varient d’un hôpital et d’un appareil à l’autre. Un classifieur entraîné dans un centre '),
        T('se dégrade sur un autre', { italics: true }),
        T(', ce qui limite le déploiement multi-sites.')
      ]),
      P([
        T('Le pipeline de référence (type '),
        T('MSclassifR', { bold: true }),
        T(') neutralise ces artefacts par un pré-traitement lourd fait main (stabilisation de variance, ondelettes, SNIP, normalisation TIC, binning 3 Da, correction de lot). '),
        T('Question de ce travail :', { bold: true, color: BLUE }),
        T(' peut-on plutôt '),
        T('apprendre l’invariance aux artefacts', { bold: true }),
        T(', à partir de spectres quasi-bruts, sans ce pré-traitement expert ?')
      ]),

      // ===== 2. APPROCHE =====
      H1('2. Approche proposée'),
      P([
        T('Un encodeur '),
        T('CNN 1D (ResNet-1D)', { bold: true }),
        T(' est pré-entraîné en '),
        T('auto-supervision VICRegL', { bold: true }),
        T(' — sans étiquettes. Les augmentations '),
        T('simulent les artefacts MALDI', { italics: true }),
        T(' (déformation de calibration m/z, ligne de base, gain, bruit, perte de pics) : le réseau apprend à ignorer ces nuisances tout en préservant la structure discriminante des pics (biomarqueurs) via le critère local de VICRegL.')
      ]),
      P([T('Entrée :', { bold: true }), T(' spectre brut ré-échantillonné (2000–20000 Da, 6000 points) + normalisation TIC seulement. ', {}),
        T('Évaluation :', { bold: true }), T(' sonde linéaire sur features gelées, comparée à un Random Forest sur '), T('binned_6000', { font: 'Consolas', size: 19 }), T(' (représentation DRIAMS standard, pré-traitement complet).')]),
      P([T('Données DRIAMS :', { bold: true }), T(' 4 centres — B (Bâle-Land), C (Aarau), D (Viollier), A (Bâle). Étiquettes '), T('Biotyper (MALDI)', { bold: true, color: ORANGE }), T('. Métrique principale : '), T('balanced accuracy', { italics: true }), T(' (classes déséquilibrées).')]),

      // ===== 3. RÉSULTATS =====
      H1('3. Résultats — vue d’ensemble'),
      makeTable(
        ['Expérience', 'Résultat clé', 'Verdict'],
        [
          { cells: ['ID espèces cross-centre B↔C', 'VICRegL 0,95–0,99 (stable) vs RF 0,68–1,00 (erratique) — ~8× plus stable', 'Favorable*'], colors: [DARK, DARK, GREEN], bold: false },
          { cells: ['Espèces proches (Enterobacter, Strepto, Klebsiella)', 'VICRegL > RF sur les 3 groupes (McNemar p<0,01)', 'Favorable*'], colors: [DARK, DARK, GREEN] },
          { cells: ['Résistance antibiotique (AMR)', 'RF bat VICRegL, aucun gain de stabilité', 'Défavorable'], colors: [DARK, DARK, RED] },
          { cells: ['Hold-out centre D (jamais vu)', 'RF 0,935 > VICRegL 0,886 — renversement', 'Défavorable'], colors: [DARK, DARK, RED] },
          { cells: ['D ajouté au pré-entraînement (non-labellisé)', '0,886 → 0,915 : écart au RF divisé par 2', 'Partiel'], colors: [DARK, DARK, ORANGE] }
        ],
        [3200, 4360, 1800]
      ),
      new Paragraph({ spacing: { before: 80, after: 120 }, children: [new TextRun({ text: '* dans le cadre où les deux centres ont été vus (non-labellisés) au pré-entraînement — voir §4.', italics: true, size: 18, color: GREY })] }),

      // ===== 4. LE POINT CRITIQUE =====
      H1('4. Le point critique : adaptation ≠ généralisation'),
      P([
        T('Le résultat marquant « ~8× plus stable » (B↔C) est obtenu quand '),
        T('les deux centres ont été vus, non-labellisés, au pré-entraînement', { bold: true }),
        T(' : c’est une '),
        T('adaptation de domaine à cible disponible', { italics: true }),
        T(', pas de la généralisation à un centre neuf. Le '),
        T('test strict', { bold: true }),
        T(' — centre D tenu totalement à l’écart du SSL — '),
        T('inverse la conclusion', { bold: true, color: RED }),
        T(' : le pipeline classique l’emporte.')
      ]),
      makeTable(
        ['Condition (éval. sur D)', 'Encodeur B∪C\n(D jamais vu)', 'Encodeur B∪C∪D\n(D vu non-labellisé)', 'RF classique'],
        [
          { cells: ['(B+C) → D  (centre inédit)', '0,886', '0,915', '0,935'], colors: [DARK, RED, ORANGE, GREEN], bold: true },
          { cells: ['Plafond D → D (sonde sur D)', '0,914', '0,931', '—'] },
          { cells: ['B+C en intra-domaine (plafond)', '0,988', '0,990', '0,889'] },
          { cells: ['AUROC « quel centre ? » sur features', '0,997', '0,999', '—'], colors: [DARK, RED, RED, DARK] }
        ],
        [3360, 2000, 2000, 2000]
      ),
      new Paragraph({ spacing: { before: 100, after: 80 }, children: [new TextRun({ text: 'Balanced accuracy ; 10 espèces communes B∩C∩D ; test = 8797 spectres de D.', italics: true, size: 18, color: GREY })] }),
      P([
        T('Diagnostic :', { bold: true, color: BLUE }),
        T(' l’AUROC « quel centre ? » reste ~1,0 même après avoir vu D — les features '),
        T('encodent encore parfaitement le centre', { bold: true }),
        T('. L’invariance revendiquée '),
        T('n’est pas atteinte', { bold: true, color: RED }),
        T(' ; ce qui fonctionnait en B↔C relevait de l’adaptation. Voir D en SSL enrichit la représentation (gain réel, sans labels) mais '),
        T('ne restaure ni la parité avec le RF ni l’invariance', { italics: true }),
        T('.')
      ]),

      // ===== 5. FORCES / LIMITES =====
      H1('5. Forces et limites'),
      H2('Forces'),
      bullet([T('Pas de pré-traitement expert', { bold: true }), T(' : spectre quasi-brut + TIC ; plafond intra-domaine élevé (0,99).')]),
      bullet([T('Extension d’un centre sans aucun label', { bold: true }), T(' : l’ajouter au SSL (non-supervisé) réduit de moitié l’écart au pipeline classique.')]),
      bullet([T('Diagnostic rigoureux', { bold: true }), T(' : décomposition plafond/transfert, classifieur de domaine, McNemar apparié, IC bootstrap.')]),
      H2('Limites'),
      bullet([T('Circularité des étiquettes', { bold: true, color: ORANGE }), T(' : DRIAMS est labellisé Biotyper (MALDI), pas NGS — 0 Shigella dans B+C. On ne peut pas prétendre « battre le MALDI » avec des labels MALDI.')]),
      bullet([T('Invariance non atteinte', { bold: true }), T(' : AUROC domaine ~1,0 ; échec sur un centre vraiment inédit.')]),
      bullet([T('AMR', { bold: true }), T(' : les augmentations (perte de pics) effacent le signal faible que la résistance requiert.')]),

      // ===== 6. CONCLUSION =====
      H1('6. Conclusion et perspectives'),
      P([
        T('Le pipeline est un '),
        T('prototype de méthode prometteur', { bold: true }),
        T(', pas un système supérieur au classique : il excelle en intra-domaine et en adaptation, mais '),
        T('l’invariance apprise ne generalise pas encore', { italics: true }),
        T(' à un centre inédit, où le pré-traitement fait main reste devant.')
      ]),
      P([T('Deux leviers pour la suite :', { bold: true, color: BLUE })]),
      bullet([T('Terme d’invariance de centre explicite', { bold: true }), T(' dans la loss (domaine-adversarial / CORAL) — attaque directement l’AUROC ~1,0, cause racine de l’échec.')]),
      bullet([T('Données à étiquettes moléculaires (WGS)', { bold: true }), T(' — seul moyen de lever la circularité et de tester la vraie hypothèse : l’invariance capte-t-elle un signal que les labels MALDI ratent ?')]),
      P([
        T('Ce dernier point ouvre une '),
        T('collaboration naturelle', { bold: true, color: GREEN }),
        T(' avec les travaux sur données MALDI appariées à des labels moléculaires (identification fine de sous-espèces, rejet par confiance).')
      ]),
      new Paragraph({ spacing: { before: 200 }, border: { top: { color: 'D5D8DC', size: 8, style: BorderStyle.SINGLE, space: 6 } }, children: [new TextRun({ text: 'Reproductibilité : encodeur ResNet-1D, PyTorch MPS (Apple M3 Pro). Résultats régénérables via scripts/08_holdout_D.py et 09_ssl_BCD.py.', italics: true, size: 17, color: GREY })] })
    ]
  }]
});

Packer.toBuffer(doc).then((buf) => {
  fs.writeFileSync('MS-VICRegL_synthese.docx', buf);
  console.log('OK -> MS-VICRegL_synthese.docx');
});
